053-configurable-event-store/proposal.md

```diff
@@ -0,0 +1,452 @@
+  pipeline. While this could be done by doing something like:
+
+  ```go
+  builds, _ := pipeline.Builds()
+  eventStore.Delete(builds)
+  ```
+
+  ...this doesn't work so well for the Postgres implementation. Currently, when
+  we delete a Pipeline from the database, `TRIGGER`s will `DROP` the
+  `pipeline_build_events_x` table for the appropriate pipeline. This results in
+  much more performant build event deletions, since dropping a table is faster
+  than deleting many thousands/millions of rows. I'd be happy to hear any clever
+  ideas anyone has on how to just have a single `Delete` method that acts on
+  builds while achieving the same table-dropping ability for Postgres.
+
+* `DeleteTeam(ctx, team db.Team)` deletes all builds for a given team. Same
+  reasoning as `DeletePipeline()`.
+
+* `UnmarshalKey(data []byte, key *Key)` works like `UnmarshalJSON` - give it
+  some bytes and tell it where you want to put the result. It is the inverse of
+  `Key.Marshal` for the corresponding `Key` type.
+
+
+## <a name="event-store-composition">`EventStore` Composition</a>
+
+In addition to enabling Concourse to interact with external stores besides
+Postgres, we can construct `EventStore` implementations that compose other
+`EventStores` to do some interesting things. One application is data
+transformations. I'll take the example of secret redaction:
+
+```go
+type SecretRedactingEventStore struct {
+    EventStore
+    // Some cred vars tracking state
+    ...
+}
+
+func (s *SecretRedactingEventStore) Initialize(ctx context.Context, build db.Build) error {
+    s.initializeCredVarsTrackingForBuild(build)
+    return s.EventStore.Initialize(ctx, build)
+}
+
+func (s *SecretRedactingEventStore) Put(ctx context.Context, build db.Build, events []atc.Event) error {
+    redactedEvents := make([]atc.Event, len(events)
+    for i, evt := range events {
+        redactedEvents[i] = s.redactSecrets(build, evt)
+    }
+    return s.EventStore.Put(ctx, build, redactedEvent)
+}
+```
+
+Another application is continuous syslog forwarding:
+
+```go
+type SyslogForwardingEventStore struct {
+    EventStore
+    // Some syslog forwarding state
+}
+
+func (s *SyslogForwardingEventStore) Put(ctx context.Context, build db.Build, events []atc.Event) error {
+    if err := s.syslogForwardEvents(build, events); err != nil {
+        return err
+    }
+    return s.EventStore.Put(ctx, build, event)
+}
+```
+
+How about migrating build events to cold-storage after they're completed:
+
+```go
+type ColdStorage interface {
+    Create(filename string) (io.WriteCloser, error)
+    Open(filename string) (io.ReadCloser, error)
+}
+
+type ColdStorageEventStore struct {
+    EventStore
+    ColdStorage ColdStorage
+}
+
+func (c *ColdStorageEventStore) Finalize(ctx context.Context, build Build) error {
+    if err := c.EventStore.Finalize(ctx, build); err != nil {
+        return err
+    }
+    // perhaps run this in the background
+    return c.migrateToColdStorage(build)
```

Comment:
I think this comment belies a lot of underlying complexity, where this design should go towards a plugin model per-provider, like Terraform (see comment below about adding providers), which uses an RPC architecture. See for example: https://www.terraform.io/docs/extend/how-terraform-works.html

This kind of multiple-binary architecture is necessary for both a) proper queuing and background processing of events as they are sent to external providers, and b) allowing additional providers to be written and developed independently of the upstream project.

Reply:
Yeah, the comment was a bit of a half-baked idea. From what I can tell, any sort of background processing we have now is just done through separate goroutines, everything residing in the same process. For instance, metric emitters run on an [`emitLoop`](https://github.com/concourse/concourse/blob/36cbf5ecf628dba92b6b4c08cfa536820638f313/atc/metric/emit.go#L136-L140) that reads from a Go `chan` that receives messages: https://github.com/concourse/concourse/blob/36cbf5ecf628dba92b6b4c08cfa536820638f313/atc/metric/emit.go#L129-L133

All that to say, we *could* (and currently do) perform some sort of background processing together without the use of plugins running in their own process and communicating via RPC. That's definitely not to say we *should* do things this way (though, admittedly, I don't fully grasp all of the implications of taking one approach over the other).